### PR TITLE
Fast current_indent (resolves #649)

### LINKED ISSUE #649: Highlight current indent (no color) with a lower priority than scope (with color)
### Problem

I just updated to version 3 today and haven't yet played around with all the options, but I couldn't find something doing what I am asking for in the following. If it already exists, could someone let me know how to do it?

I really like the new highlighting via scope (especially in Rust). Even in Lua, I think this is helpful:
Screenshot 2023-09-28 at 23 34 33

So I plan to keep the highlighting on (I am matching it with rainbow-delimiters).

But whenever there is no scope highlighting I would like to fall back to slight highlighting of my current indentation level. I.e., in this picture:
Screenshot 2023-09-28 at 23 34 56

I would like to have some slight highlighting of the second indent line to the left of my cursor. 

Preferably I would probably want to always have a slight highlighting of the current indent level unless that highlighting is overwritten by the scope highlighting. So sometimes the might be two highlighted indent lines at the same time. E.g., here:

Screenshot 2023-09-28 at 23 46 15

I would like to keep the first indent red, but have the second indent slightly light up to indicate the current indent level at the cursor position. 

Is there already a way to do this? If not, I am willing to spend some time to try helping with implementation or testing of something like that. And if nothing else, I would like to request a feature like that. 

### Expected behavior

Always slightly highlight (e.g., just brighter color) the current indent level unless the current indent level is also the scope level in which case the scope is preferred for highlighting. 

(Alternatively, when not using the current scope to highlight anything, highlight slightly the current indent level.)

### PR DESCRIPTION
**Pull request by Danielkonge**
Here is a cleaned up version of my implementation of `current_indent` that doesn't contain other features.

This pull request will resolve #649.

### CODE REVIEW

**Comment by Danielkonge** [comment]
@lukas-reineke I think this should be basically ready now, but I cannot figure out what stylua is complaining about? I think it might just not like the `::continue::` syntax, but it weirdly accepts the later one. Can you tell what is wrong?

File: lua/ibl/init.lua [hunk 1]
```diff
@@ -365,13 +418,78 @@ M.refresh = function(bufnr)
...
-        last_whitespace_tbl = whitespace_tbl
+        -- this do block is purely to help stylua with the scope
+        do
+            arr_whitespace_tbl[i] = whitespace_tbl
+        end
+
+        ::continue::
+    end
+
+    -- between the two loops we find the current indent column and current indent start quickly
```
Comment (lukas-reineke) [inline lua/ibl/init.lua:429]: Why does this need to happen between the loop? If you keep track of the current indent in the first loop, you can just use that as the current indent when you hit the current line

**Review by lukas-reineke — CHANGES_REQUESTED** [review]
Current indent should also support `show_start` and `show_end`.
And you need to add `priority` that decides if scope or current indent is displayed if they are in the same place.

**Comment by lukas-reineke** [comment]
I get this behavior with tabs. The second if block should not be highlighted

File: lua/ibl/init.lua (hunk 1 above)
Comment (Danielkonge) [reply lua/ibl/init.lua:429]: I have updated this now so it happens in the loop.

**Comment by Danielkonge** [comment]
> I get this behavior with tabs. The second if block should not be highlighted
> 
> 

This should be fixed now. (Please let me know if it not.) I used `indent_state.stack` to decide what indents were drawn, but I can see it doesn't work as I would expect in all cases (e.g. with the blank lines in the if statements like in the picture). To overcome this I have constructed a table `drawn_indents` alongside the `whitespace_tbl` that keeps track of exactly which indents have been drawn. This also simplifies a workaround with blank lines I used earlier.

**Comment by Danielkonge** [comment]
I have added `priority`, `show_start` and `show_end` now. Note that `show_end` doesn't really make sense in the same way for the indentation guides as for scope, so the only thing I think makes sense here is to underline the line below the end of the current indent level if you enable it.

**Comment by lukas-reineke** [comment]
I don't quite understand the need for `drawn_indents`, isn't it just the same information you already have in `whitespace_tbl`?

**Comment by Danielkonge** [comment]
> I don't quite understand the need for `drawn_indents`, isn't it just the same information you already have in `whitespace_tbl`?

The information in `drawn_indents` is already in `whitespace_tbl`, yes. I just thought it was easier to work with a list of purely the indents, whereas `whitespace_tbl` includes spaces and other stuff too. If you prefer, I can rewrite it to just use `whitespace_tbl` directly (you will just need to search backwards through `whitespace_tbl` for the indents).

`drawn_indents` just contains the positions of `TAB_START`, `TAB_START_SINGLE` and `INDENT` in the `whitespace_tbl` (and it ignores `TAB_FILL`, `TAB_END` and `SPACE`), since these are the ones used for drawing the indentation guides.

**Comment by lukas-reineke** [comment]
👍 yeah that's what I mean. Let's just use `whitespace_tbl`. You can make a helper function to give you back the indents.
It shouldn't hurt performance, and it's easier to understand when there is only one data source.

**Comment by Danielkonge** [comment]
> 👍 yeah that's what I mean. Let's just use `whitespace_tbl`. You can make a helper function to give you back the indents. It shouldn't hurt performance, and it's easier to understand when there is only one data source.

Since I only actually really need the last indent column and the number of indents I have made a helper function to return these given a whitespace table now.

**Comment by Danielkonge** [comment]
Since the other pull request is almost ready, I will close this one.

**Comment by wookayin** [comment]
Continued in #743 (adding a ref here because they were not linked). Thanks @Danielkonge for the great work.

**Pull request closed without merging**